### ISSUE TITLE
`__res_init: symbol not found` when running Prisma

### ISSUE THREAD

**Issue by cmwhited** [prisma/prisma#14043]
> What error exactly @cmwhited? Can you try with a `prisma@dev` and `@prisma/client@dev` (or version 4.0.0 which will be release tomorrow) please and post the full output you are getting? That would help us to make sure you ar egetting the same error indeed.

Updated to v4.0.0 and it still fails. I think it is actually an issue with `pnpm`. Testing now

### update

after upgrading to v4, it is now throwing this more verbose error
```
#21 2.546 Error: Get Config: Unable to establish a connection to query-engine-node-api library. It seems there is a problem with your OpenSSL installation!
#21 2.546 Details: Unable to require(`/app/node_modules/.pnpm/prisma@4.0.0/node_modules/prisma/libquery_engine-linux-arm64-openssl-1.1.x.so.node`)
#21 2.546  Error relocating /app/node_modules/.pnpm/prisma@4.0.0/node_modules/prisma/libquery_engine-linux-arm64-openssl-1.1.x.so.node: __res_init: symbol not found
```

### update 2

Switching from `pnpm` to `yarn` does not solve the issue, throws error

```
#18 1.179 api:gen:types: Prisma schema loaded from prisma/schema.prisma
#18 1.419 api:gen:types: Error: Get Config: Unable to establish a connection to query-engine-node-api library. It seems there is a problem with your OpenSSL installation!
#18 1.421 api:gen:types: Details: Unable to require(`/app/node_modules/prisma/libquery_engine-linux-arm64-openssl-1.1.x.so.node`)
#18 1.421 api:gen:types:  Error relocating /app/node_modules/prisma/libquery_engine-linux-arm64-openssl-1.1.x.so.node: __res_init: symbol not found
#18 1.421 api:gen:types: 
#18 1.421 api:gen:types: Prisma CLI Version : 4.0.0
#18 1.438 api:gen:types: error Command failed with exit code 1.
```

idk if it helps, but here are my versions
- nodejs: v16.15.0
- yarn: v1.22.17
- pnpm: v7.3.0
- mac osx: v12.4, m1 chip
- docker: version 20.10.16, build aa7e414

_Originally posted by @cmwhited in https://github.com/prisma/prisma/issues/13717#issuecomment-1169255951_

**Comment by cmwhited** [comment]
> Follow up question 1 (please reply in new issue): Does `/app/node_modules/prisma/libquery_engine-linux-arm64-openssl-1.1.x.so.node` exist?

it does not exist @janpio 

<img width="495" alt="Screen Shot 2022-06-28 at 12 30 59 PM" src="https://user-images.githubusercontent.com/18075124/176315285-92b87b53-2a03-4ffc-9108-13bcc12b66cc.png">

**Comment by janpio** [comment]
Earlier, additional information from @cmwhited on his system: https://github.com/prisma/prisma/issues/13717#issuecomment-1167542431

**Comment by jkomyno** [comment]
This is probably a duplicate of https://github.com/prisma/prisma/issues/9572. More details will follow

**Comment by cmwhited** [comment]
@jkomyno @janpio I went down the path as referenced here: https://github.com/prisma/prisma/issues/8478#issuecomment-1097685610

and updated my `Dockerfile` to

```
FROM positivly/prisma-binaries:latest as prisma
FROM node:16.15-alpine3.16
RUN apk add --no-cache git curl python3 make gcc libc-dev g++ libc6-compat openssl openssl-dev \
    && curl -sL https://unpkg.com/@pnpm/self-installer | node
RUN ln -sf python3 /usr/bin/python
COPY --from=prisma /prisma-engines/query-engine /prisma-engines/migration-engine /prisma-engines/
ENV PRISMA_QUERY_ENGINE_BINARY=/prisma-engines/query-engine \
  PRISMA_MIGRATION_ENGINE_BINARY=/prisma-engines/migration-engine \
  PRISMA_CLI_QUERY_ENGINE_TYPE=binary \
  PRISMA_CLIENT_ENGINE_TYPE=binary

WORKDIR /app

EXPOSE 4000/tcp

ARG DATABASE_URL

ENV DATABASE_URL=${DATABASE_URL}

COPY ./package.json /app/package.json
COPY ./pnpm-workspace.yaml /app/pnpm-workspace.yaml
COPY ./turbo.json   /app/turbo.json
COPY ./apps/api/package.json  /app/apps/api/package.json
COPY ./apps/api/prisma/schema.prisma /app/apps/api/prisma/schema.prisma
COPY ./pnpm-lock.yaml /app/pnpm-lock.yaml

# install the packages

COPY . .

# build .env in /app/apps/api

RUN cd /app/apps/api \
  && touch .env \
  && echo "DATABASE_URL=${DATABASE_URL}" >> .env \
  && cat .env

RUN --mount=type=secret,id=npmrc,dst=/root/.npmrc \
  pnpm i --prod false

RUN pnpm gen:types:api

RUN pnpm build
```

This fails with a lovely new error:

```
#23 1.132 api:gen:types: > prisma generate
#23 1.132 api:gen:types: 
#23 1.574 api:gen:types: Environment variables loaded from .env
#23 1.576 api:gen:types: Prisma schema loaded from prisma/schema.prisma
#23 1.946 api:gen:types: Error: 
#23 1.946 api:gen:types: TypeError: collection is not iterable
#23 1.946 api:gen:types:     at keyBy (/app/node_modules/.pnpm/@prisma+client@4.0.0_prisma@4.0.0/node_modules/@prisma/client/generator-build/index.js:60084:21)
#23 1.946 api:gen:types:     at DMMFDatamodelHelper.getTypeMap (/app/node_modules/.pnpm/@prisma+client@4.0.0_prisma@4.0.0/node_modules/@prisma/client/generator-build/index.js:61890:17)
#23 1.946 api:gen:types:     at new DMMFDatamodelHelper (/app/node_modules/.pnpm/@prisma+client@4.0.0_prisma@4.0.0/node_modules/@prisma/client/generator-build/index.js:61880:25)
#23 1.946 api:gen:types:     at new BaseDMMFHelper (/app/node_modules/.pnpm/@prisma+client@4.0.0_prisma@4.0.0/node_modules/@prisma/client/generator-build/index.js:62031:32)
#23 1.946 api:gen:types:     at new DMMFHelper (/app/node_modules/.pnpm/@prisma+client@4.0.0_prisma@4.0.0/node_modules/@prisma/client/generator-build/index.js:62038:32)
#23 1.946 api:gen:types:     at new TSClient (/app/node_modules/.pnpm/@prisma+client@4.0.0_prisma@4.0.0/node_modules/@prisma/client/generator-build/index.js:63260:17)
#23 1.946 api:gen:types:     at buildClient (/app/node_modules/.pnpm/@prisma+client@4.0.0_prisma@4.0.0/node_modules/@prisma/client/generator-build/index.js:63525:24)
#23 1.946 api:gen:types:     at generateClient (/app/node_modules/.pnpm/@prisma+client@4.0.0_prisma@4.0.0/node_modules/@prisma/client/generator-build/index.js:63592:47)
#23 1.946 api:gen:types:     at async LineStream.<anonymous> (/app/node_modules/.pnpm/@prisma+client@4.0.0_prisma@4.0.0/node_modules/@prisma/client/generator-build/index.js:56452:24)
#23 1.946 api:gen:types: 
#23 1.946 api:gen:types: 
#23 1.961 api:gen:types:  ELIFECYCLE  Command failed with exit code 1.
#23 1.976 api:gen:types: ERROR: command finished with error: command (apps/api) pnpm run gen:types exited (1)
#23 1.976 command (apps/api) pnpm run gen:types exited (1)
```

Going to issue https://github.com/prisma/prisma/issues/9572 lead me to there

**Comment by janpio** [comment]
Quick side note: Our error message `It seems there is a problem with your OpenSSL installation!` is definitely wrong and misleading here. The problem is most probably that we do not have the correct  engine file for your environment, or Prisma does not correctly detect that, see: https://github.com/prisma/prisma/issues/8478

**Comment by cmwhited** [comment]
> Quick side note @cmwhited: Our error message `It seems there is a problem with your OpenSSL installation!` is definitely wrong and misleading here. The problem is most probably that we do not have the correct engine file for your environment, or Prisma does not correctly detect that, see: #8478

I added: 
```
binaryTargets = ["native", "linux-arm64-openssl-1.1.x"]
```
to the `schema.prisma` which I can see generates the correct engine file, but I still get the 
```
#23 1.576 api:gen:types: Prisma schema loaded from prisma/schema.prisma
#23 1.946 api:gen:types: Error: 
#23 1.946 api:gen:types: TypeError: collection is not iterable
#23 1.946 api:gen:types:     at keyBy (/app/node_modules/.pnpm/@prisma+client@4.0.0_prisma@4.0.0/node_modules/@prisma/client/generator-build/index.js:60084:21)
#23 1.946 api:gen:types:     at DMMFDatamodelHelper.getTypeMap (/app/node_modules/.pnpm/@prisma+client@4.0.0_prisma@4.0.0/node_modules/@prisma/client/generator-build/index.js:61890:17)
#23 1.946 api:gen:types:     at new DMMFDatamodelHelper (/app/node_modules/.pnpm/@prisma+client@4.0.0_prisma@4.0.0/node_modules/@prisma/client/generator-build/index.js:61880:25)
#23 1.946 api:gen:types:     at new BaseDMMFHelper (/app/node_modules/.pnpm/@prisma+client@4.0.0_prisma@4.0.0/node_modules/@prisma/client/generator-build/index.js:62031:32)
#23 1.946 api:gen:types:     at new DMMFHelper (/app/node_modules/.pnpm/@prisma+client@4.0.0_prisma@4.0.0/node_modules/@prisma/client/generator-build/index.js:62038:32)
#23 1.946 api:gen:types:     at new TSClient (/app/node_modules/.pnpm/@prisma+client@4.0.0_prisma@4.0.0/node_modules/@prisma/client/generator-build/index.js:63260:17)
#23 1.946 api:gen:types:     at buildClient (/app/node_modules/.pnpm/@prisma+client@4.0.0_prisma@4.0.0/node_modules/@prisma/client/generator-build/index.js:63525:24)
#23 1.946 api:gen:types:     at generateClient (/app/node_modules/.pnpm/@prisma+client@4.0.0_prisma@4.0.0/node_modules/@prisma/client/generator-build/index.js:63592:47)
#23 1.946 api:gen:types:     at async LineStream.<anonymous> (/app/node_modules/.pnpm/@prisma+client@4.0.0_prisma@4.0.0/node_modules/@prisma/client/generator-build/index.js:56452:24)
#23 1.946 api:gen:types: 
```
error

**Comment by cmwhited** [comment]
@janpio switching docker to the `16.15-slim` image _does_ work. So it is an `xx-alpine` docker image issue

**Comment by TeemuKoivisto** [comment]
> @janpio switching docker to the `16.15-slim` image _does_ work. So it is an `xx-alpine` docker image issue

Yeah somebody should definitely put up somewhere that installing Prisma on alpine crashes with these most bizarre error messages. Switching to slim and adding `RUN apt-get update && apt-get -y install openssl` fixed this for me.

**Comment by bmxpiku** [comment]
I'm using `node:18-alipine`, upgraded from prisma `3.14` to `4.3.1` and now I experience the same issue...

**Comment by nassim-kachroud** [comment]
Same issue on my side using prisma 4.x.x, any update?

**Comment by DavorJ94** [comment]
@nassim-kachroud @bmxpiku maybe try with this in your alpine container as soon as you pull initial image, it solved for me:

```RUN apk upgrade --update-cache --available && \
    apk add openssl && \
    apk add openssl-dev && \
    rm -rf /var/cache/apk/*```

**Comment by vuong-nguyen-saltovn** [comment]
```
$ prisma generate
Environment variables loaded from .env
Prisma schema loaded from prisma/schema.prisma
Error: Unable to require(`/backend/node_modules/prisma/libquery_engine-linux-musl.so.node`)
 Error loading shared library libssl.so.1.1: No such file or directory (needed by /backend/node_modules/prisma/libquery_engine-linux-musl.so.node)
 ```
I got this error when build in docker, too.

**Comment by senorscan** [comment]
Same issue, I tried with your suggestion @DavorJ94, however still no luck. Tried `16.15-slim` and it's working. Must have been some regression in the latest version of `16-alpine`

**Comment by DavorJ94** [comment]
@senorscan Interesting, it works for me for 16.12-alpine, but for 18 it won't work. It seems there really is some regression. Maybe you can try 16.12.

**Comment by vuong-nguyen-saltovn** [comment]
> @senorscan Interesting, it works for me for 16.12-alpine, but for 18 it won't work. It seems there really is some regression. Maybe you can try 16.12.

It worked! Thank you very much.

**Comment by zackdotcomputer** [comment]
Thanks @edlundin - that unblocked me. Adding `RUN apk add --update --no-cache openssl1.1-compat` to my dockerfile has fixed builds without needing to resort to the downgrade/crossgrade/emulation solutions on other threads.

**Comment by valentinpezon-primo** [comment]
Hey, just had this exact issue, thanks for the solution.

@zackdotcomputer, so you added `RUN apk add --update --no-cache openssl1.1-compat` and kept `node:lts-alpine` ?

**Comment by zackdotcomputer** [comment]
@valentinpezon-primo In my case I'm on `node:fermium-alpine` still but yes to the rest. Just needed to be sure to add the `apk` command before the `npm install` command because of the prisma post-install hook.

**Comment by DavorJ94** [comment]
@valentinpezon-primo yes, I just tried it with node:18.12-alpine, it seems to work for any version. Perfect, thanks.

**Comment by maitrungduc1410** [comment]
weird, it was working fine, and just from this morning it has been failing. no change from my Dockerfile at all

**Comment by ugurkellecioglu** [comment]
I specified node version as node:alpine in my Dockerfile, I switched it to 16.2 and fixed.

**Comment by rained23** [comment]
getting this issue in my bitbucket pipelines too

**Comment by Maetes** [comment]
When I add

> weird, it was working fine for months, and just from this morning it has been failing in my Gitlab CI, building manually from local is fine. No change from my Dockerfile at all.
> 
> Adding this before `npm install` fixed the issue:
> 
> ```dockerfile
> RUN apk add --update --no-cache openssl1.1-compat
> ```

When I Add this line I can build the Docker-file but I get the error:

```
In case this error is unexpected for you, please report it in https://github.com/prisma/prisma/issues
Error: @prisma/client did not initialize yet. Please run "prisma generate" and try to import it again.
In case this error is unexpected for you, please report it in https://github.com/prisma/prisma/issues
```

Its strange cause in my Dockerfile I copy the whole node_modules folder to the build step...:

```
FROM node:18-alpine as development
RUN apk add --update --no-cache openssl1.1-compat
WORKDIR /usr/src/app
COPY --chown=node:node package*.json ./
RUN npm ci
COPY --chown=node:node . .
RUN npm run prisma:generate
USER node

FROM node:18-alpine AS builder
WORKDIR /usr/src/app
COPY --chown=node:node package*.json ./
COPY --chown=node:node --from=development /usr/src/app/node_modules ./node_modules
COPY --chown=node:node . .
RUN npm run build
ENV NODE_ENV production
RUN npm ci --only=production && npm cache clean --force
USER node

...

```

**Comment by Geczy** [comment]
fixed as well by switching to `FROM node:16.2.0-alpine AS deps`

**Comment by jkomyno** [comment]
Hey, Alberto from Prisma here.

For people facing this issue on Linux Alpine (e.g., when using the `node:alpine`, `node:lts-alpine` Docker images), please refer to this solution: https://github.com/prisma/prisma/issues/16553#issuecomment-1353302617

**Comment by jkomyno** [comment]
Closing as the issue is solved by https://github.com/prisma/prisma/issues/14043#issuecomment-1353364540
